dlang install: Fully automate installation of DMD

leandro-lucarella-sociomantic opened the pull request:
The `dlang install` command now uses a much more automated approach to building the image, by using a `beaver.Dockerfile` and `docker/build` script. It will pass the `DMD_PKG` argument to `docker build` install the specified package.

This also moves the tests for this command to the single-tests framework.
    
Fixes #13.

leandro-lucarella-sociomantic committed:
Add missing standard .dockerignore

Even when there is no docker directory for now, we leave the exception.

leandro-lucarella-sociomantic committed:
docker run: Do not pass -ti by default

Even when running docker manually and interactively one usually pass -ti,
for running tests this is not usually desired, specially if the output of
docker run needs to be captured (see
https://github.com/moby/moby/issues/8513 for more details on this issue).

leandro-lucarella-sociomantic committed:
Add auto-discovered single tests

It doesn't scale well to run all the tests right from the .travis.yml file.
Is good to test the basic layout of a .travis.yml for a beaver project but
beyond that, tests can live separately and just ran once in Travis.

This adds a new matric cell to just run these tests, separately from the
travis matrix tests.

leandro-lucarella-sociomantic committed:
docker: Add new gen-dockerfile command

The `docker` command learned a new subcommand to generate a `Dockerfile`
based on a `beaver.Dockerfile` (by default). Please look read `beaver
docker gen-dockerfile -h` for more options and information on how the
file is actually generated.

This commit also includes tests using the new single-tests framework.

Part of #13.

leandro-lucarella-sociomantic commented:
The <del>first</del> second (by github failed visualization of authorship date instead of commit date) commit is actually a bug fix, but it's a bit tricky and since v0.2.0 will be out soon and we don't really have all that many users of beaver yet (and we are at v0.x still!), I prefer to include it in the next non-patch release. I want v0.1.x to die soon! :gun:

nemanja-boric-sociomantic commented on bin/docker/gen-dockerfile:
```diff
@@ -0,0 +1,117 @@
+set -eu
+
+# Defaults
+default_input=Dockerfile.beaver
```
`Dockerfile.inc`?

leandro-lucarella-sociomantic commented:
BTW, next step is to integrate this with `beaver dlang install`, to get all the `ARG` stuff sorted out...

nemanja-boric-sociomantic commented on bin/docker/gen-dockerfile:
```diff
@@ -0,0 +1,117 @@
+docker_build=build
+docker_dist=${DIST:-$(lsb_release -cs)}
+
+# Transforms a fragment (for now it just injects the DIST
```
unbalanced parenthesis

nemanja-boric-sociomantic replied:
The entire comment looks cropped.

leandro-lucarella-sociomantic replied:
No, actually the commit message / relnotes are wrong... Using `.inc` makes VIM not syntax highlight properly, as it thinks the file is something else. Actually I should probably use `beaver.Dockerfile`, I think with that extension it should never fail :)

nemanja-boric-sociomantic replied:
I think this reply from you was referring to the name of `Dockerfile.inc` and now github somehow merged it with this comment. You're still missing the closing `)`.

leandro-lucarella-sociomantic commented:
Updated (wow! It was green already! :D)

leandro-lucarella-sociomantic committed:
Add a local-test file to ease running test locally

Testing using Travis has a very slow turnaround and adds unfair stress to
Travis servers. This script can be used to trigger most of the tests
locally.

leandro-lucarella-sociomantic committed:
install: Use gen-dockerfile if beaver.Dockerfile is present

The `beaver install` command learned to use the `beaver.Dockerfile` to
generate the `Dockerfile`. The `Dockerfile.$DIST` still has precedence
over the `beaver.Dockerfile` in case customization of `Dockerfile` in
a way that `beaver docker gen-dockerfile` is not enough is **really**
needed.

Another advance towards #13.

leandro-lucarella-sociomantic commented:
> BTW, next step is to integrate this with beaver dlang install, to get all the ARG stuff sorted out...

Now plugged into `beaver install`, one last step to get it integrated in dlang builds to have a super simplified procedure!

leandro-lucarella-sociomantic committed:
dlang: Make install install DMD automatically

The `dlang install` command now uses a much more automated approach to
building the image, by using a `beaver.Dockerfile` and `docker/build`
script (like `beaver install`). It will pass the `DMD_PKG` argument to
`docker build` install the specified package.

This also moves the tests for this command to the single-tests framework.

Fixes #13.

leandro-lucarella-sociomantic committed:
dlang install: Install a matching phobos

When `dmd` is used, when installing the `dmd-bin` package we also need to
specify the `libphobos2-dev` package version.

leandro-lucarella-sociomantic committed:
Add a simple Makefile for easy local testing

It just runs the local-test script, but now that we don't have
a Makefile that is needed for the tests themselves, is good to have
a consistent `make test` in all projects.

leandro-lucarella-sociomantic commented:
OK, I consider this finished and ready for a final review. To be honest, I'm quite happy with the results!

I will update the [ocean PR](https://github.com/sociomantic-tsunami/ocean/pull/209) and [swarm PR](https://github.com/sociomantic-tsunami/swarm/pull/103) to use this new feature for a final test.

leandro-lucarella-sociomantic commented:
Ocean passes :)

leandro-lucarella-sociomantic commented:
Swarm too! :D

nemanja-boric-sociomantic commented:
OK, I'm going to give this a full last review then.

nemanja-boric-sociomantic commented on .travis.yml:
```diff
@@ -23,13 +23,15 @@ env:
-    - beaver dlang make
+    - test -n "$SINGLE" || beaver docker run test/hello
+    - test -n "$SINGLE" || beaver dlang make
+    - if test -n "$SINGLE"; then test/single-tests; fi
```
`test -n "$SINGLE" && test/single-tests` to match the style of others?

leandro-lucarella-sociomantic replied:
I'm used to not use `&&` because if `sh -e` is set, it fill stop and make the script fail just when the first test fails. But in this case it should be the same because is the last command in the script. If the `test` fails, then travis will see the whole script failing and will mark the build as failed.

leandro-lucarella-sociomantic replied:
Maybe I should add a comment and/or use `if` everywhere...

mihails-strasuns-sociomantic replied:
Yeah I'd prefer to have single condition style too, whatever it is.

mihails-strasuns-sociomantic commented on bin/dlang/install:
```diff
@@ -12,7 +12,7 @@ case "$DMD" in
     2.*.s* ) PKG="dmd-transitional=$DMD-$DIST" ;;
-    2.*    ) PKG="dmd-bin=$DMD" ;;
+    2.*    ) PKG="dmd-bin=$DMD libphobos2-dev=$DMD" ;;
     *      ) echo "Unknown \$DMD ($DMD)" >&2; false ;;
```
What is the purpose of plain `false` here? Should't it be `exit something`?

leandro-lucarella-sociomantic replied:
Since `set -e` is used, it will make the script fail, but yeah, a explicit `exit` could be used too.

nemanja-boric-sociomantic commented on bin/docker/gen-dockerfile:
```diff
@@ -0,0 +1,121 @@
+do
+	case "$arg" in
+        i) fragments+=("$(cat "$OPTARG")") ;;
+        I) fragments+=("$OPTARG") ;;
```
indentation

leandro-lucarella-sociomantic replied:
o_O

leandro-lucarella-sociomantic replied:
I have no idea why I have mixed spaces and tabs in there. Will fix.

nemanja-boric-sociomantic commented on local-test:
```diff
@@ -0,0 +1,26 @@
+    shift
+    "./$t"
+    echo
+    cd -
```
I saw `cd -` somewhere else as well. Would it work if test `cd`s internally? I wish `sh` had `pushd/popd`.

leandro-lucarella-sociomantic replied:
Yes. The script runs in a subshell that won't affect the parent's shell current working directory.

mihails-strasuns-sociomantic commented on test/docker/gen-dockerfile/beaver.Dockerfile:
```diff
@@ -0,0 +1,8 @@
+#
+# This Dockerfile is just for testing
+FROM sociomantictsunami/dlang:v2
+RUN touch /BUILT
```
Is this intended to touch same file as the relevant `build` script?

leandro-lucarella-sociomantic replied:
Nope, it must be a leftover from some tests redo :disappointed:

nemanja-boric-sociomantic commented on relnotes/install-gen-dockerfile.feature.md:
````diff
@@ -0,0 +1,47 @@
+  ```
+
+  You can completely omit the `docker/build` script if you don't need any extra
+  build steps, if the script doesn't exist, nothing will be copied to the image
````
suggestion: `build steps - if the script doesn't exists, nothing ...` (dash instead of comma) Also, maybe this point should be mentioned in `README`?

leandro-lucarella-sociomantic replied:
Maybe just removing the second "comma" in my original sentence? Invoking @gavin-norman-sociomantic powers!

gavin-norman-sociomantic replied:
Yeah, I'd suggest a `;` or `-` in place of the comma here "build steps, if the script doesn't exist".

leandro-lucarella-sociomantic replied:
:+1:

nemanja-boric-sociomantic commented on README.md:
````diff
@@ -172,6 +172,10 @@ else
 fi
 ```
 
+You can completely omit the `docker/build` script if you don't need any extra
````
Ah, ok. I suggested this in the previous commit (where it should be included I think).

leandro-lucarella-sociomantic replied:
I messed up a commit, a readme that should have been amended in a commit was done in the wrong one. I will fix the commits.

nemanja-boric-sociomantic commented:
Apart from the small issues, LGTM. I'll assign @mihails-strasuns-sociomantic .

leandro-lucarella-sociomantic commented:
Updated to address all the comments. Let me know if there is anything else missing.

Here is a diff against the previous state for convenience:

~~~diff
diff --git a/.travis.yml b/.travis.yml
index 259c61e..c22a6ea 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -28,9 +28,18 @@ env:
         - DIST=trusty DMD=dmd1
         - DIST=xenial DMD=2.070.2.s12
 
-install:
-    - test -n "$SINGLE" || beaver install
+install: |
+    set -e
+    if test -z "$SINGLE"
+    then
+        beaver install
+    fi
 
-script:
-    - test -n "$SINGLE" || beaver docker run test/hello
-    - if test -n "$SINGLE"; then test/single-tests; fi
+script: |
+    set -e
+    if test -z "$SINGLE"
+    then
+        beaver docker run test/hello
+    else
+        test/single-tests
+    fi
diff --git a/README.md b/README.md
index 0e1a27b..c4e5351 100644
--- a/README.md
+++ b/README.md
@@ -173,7 +173,7 @@ fi
 ```
 
 You can completely omit the `docker/build` script if you don't need any extra
-build steps, if the script doesn't exist, nothing will be copied to the image
+build steps - if the script doesn't exist, nothing will be copied to the image
 building procedure either.
 
 The `Dockerfile` generation is performed by the `beaver docker gen-dockerfile`
diff --git a/bin/dlang/install b/bin/dlang/install
index 8c3398b..f82a80f 100755
--- a/bin/dlang/install
+++ b/bin/dlang/install
@@ -15,7 +15,7 @@ case "$DMD" in
     1.*    ) PKG="dmd1=$DMD-$DIST" ;;
     2.*.s* ) PKG="dmd-transitional=$DMD-$DIST" ;;
     2.*    ) PKG="dmd-bin=$DMD libphobos2-dev=$DMD" ;;
-    *      ) echo "Unknown \$DMD ($DMD)" >&2; false ;;
+    *      ) echo "Unknown \$DMD ($DMD)" >&2; exit 1 ;;
 esac
 
 # Generate the Dockerfile including the DMD_PKG argument and install the
diff --git a/bin/docker/gen-dockerfile b/bin/docker/gen-dockerfile
index 8cb4993..99303c6 100755
--- a/bin/docker/gen-dockerfile
+++ b/bin/docker/gen-dockerfile
@@ -16,7 +16,7 @@ docker_dist=${DIST:-$(lsb_release -cs)}
 # Command usage string / help
 usage()
 {
-	cat <<EOT
+    cat <<EOT
 Usage: $0 [options]
 
 Generate a Dockerfile based on a Dockerfile fragments. By default the file
@@ -65,16 +65,16 @@ declare -a fragments
 output=
 while getopts i:I:o:d:b:D:h arg
 do
-	case "$arg" in
+    case "$arg" in
         i) fragments+=("$(cat "$OPTARG")") ;;
         I) fragments+=("$OPTARG") ;;
-		o) output="$OPTARG" ;;
-		d) docker_dir="$OPTARG" ;;
-		b) docker_build="$OPTARG" ;;
-		D) docker_dist="$OPTARG" ;;
-		h) usage ; exit 0 ;;
-		\?) usage >&2; exit 2 ;;
-	esac
+        o) output="$OPTARG" ;;
+        d) docker_dir="$OPTARG" ;;
+        b) docker_build="$OPTARG" ;;
+        D) docker_dist="$OPTARG" ;;
+        h) usage ; exit 0 ;;
+        \?) usage >&2; exit 2 ;;
+    esac
 done
 shift $(($OPTIND-1))
 
diff --git a/relnotes/install-gen-dockerfile.feature.md b/relnotes/install-gen-dockerfile.feature.md
index 9ed7c30..073b60b 100644
--- a/relnotes/install-gen-dockerfile.feature.md
+++ b/relnotes/install-gen-dockerfile.feature.md
@@ -43,5 +43,5 @@
   \```
 
   You can completely omit the `docker/build` script if you don't need any extra
-  build steps, if the script doesn't exist, nothing will be copied to the image
+  build steps - if the script doesn't exist, nothing will be copied to the image
   building procedure either.
diff --git a/test/docker/gen-dockerfile/docker/build b/test/docker/gen-dockerfile/docker/build
index e8336f2..f7df550 100755
--- a/test/docker/gen-dockerfile/docker/build
+++ b/test/docker/gen-dockerfile/docker/build
@@ -4,4 +4,4 @@
 # (See accompanying file LICENSE.txt or copy at
 # http://www.boost.org/LICENSE_1_0.txt)
 
-touch /BUILT
+touch /BUILT-build
diff --git a/test/docker/gen-dockerfile/test b/test/docker/gen-dockerfile/test
index 76f4990..0714ec9 100755
--- a/test/docker/gen-dockerfile/test
+++ b/test/docker/gen-dockerfile/test
@@ -14,7 +14,7 @@ cat Dockerfile
 grep -q "FROM.*:$(lsb_release -cs)-v2" Dockerfile # Check for the host's DIST
 beaver docker build .
 rm Dockerfile
-beaver docker run test -f /BUILT
+beaver docker run test -f /BUILT -a -f /BUILT-build
 test "$(beaver docker run lsb_release -cs)" = "$(lsb_release -cs)"
 
 # Test basic usage but with DIST env var
@@ -26,7 +26,7 @@ test_basic_dist()
     grep -q "FROM.*:$DIST-v2" Dockerfile2
     beaver docker build -f Dockerfile2 .
     rm Dockerfile2
-    beaver docker run test -f /BUILT
+    beaver docker run test -f /BUILT -a -f /BUILT-build
     test "$(beaver docker run lsb_release -cs)" = "$DIST"
 }
 test_basic_dist trusty
~~~

mihails-strasuns-sociomantic commented:
@leandro-lucarella-sociomantic naiming of touched files in test scripts still looks like a mess with uncomprehensible relationship. Maybe you can unify them all under one naming scheme?

leandro-lucarella-sociomantic commented:
Better now? If not please propose some naming scheme that you think it would make it clearer.

mihails-strasuns-sociomantic reviewed (approved):
Yeah, LGTM now!
